MAINT Use check_scalar in BaseDecisionTree

genvalen opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Addresses #20724
#DataUmbrella

#### What does this implement/fix? Explain your changes.
Summary of changes to `BaseDecisionTree`:

* Add tests to ensure estimator raises proper errors when invalid arguments are passed in. 
* Use the helper function `check_scalar` from `sklearn.utils` to validate the scalar parameters.

Test and validation progress:
- [x] max_depth
- [x] min_samples_split
- [x] min_samples_leaf
- [x] min_weight_fraction_leaf
- [x] max_features
- [x] max_leaf_nodes
- [x] min_impurity_decrease
- [x] ccp_alpha

References
1. [check_scalar docs](https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_scalar.html)
2. [PR #20723](https://github.com/scikit-learn/scikit-learn/pull/20723)

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

genvalen committed:
max_depth: add tests

genvalen committed:
max_depth: add validation

genvalen committed:
min_sample_split: add tests and validation

genvalen committed:
min_impurity_decrease: add tests and valiation

genvalen committed:
fix lint issue

genvalen committed:
max_leaf_nodes: add tests and validation

genvalen commented:
**Notes on param range in trees**
(param: type, range)
`max_depth`: int, if not none, then [1, ) 
`min_samples_split`: int -> [2, number samples], float-> (0, 1)
`min_samples_leaf`: int -> [1, number of samples], float-> (0, 1)
`min_weight_fraction_leaf`: float-> [0, 1)
`max_features`: int -> [1, number of features], float (0, 1), or string (not checked with check_scalar)
`max_leaf_nodes`: int, if not none, then [1, )
`min_impurity_decrease`: float-> [0, )

genvalen committed:
min_weight_fraction_leaf: add tests and validation

genvalen committed:
format files w/ black

genvalen committed:
remove comments

genvalen commented:
Hi @glemaitre, the range for these parameters is different in the existing code and in my notes:
        `min_samples_leaf`:  float-> (0, 0.5] vs (0, 1) 
        `min_weight_fraction_leaf`: float-> (0, 0.5] vs [0, 1)
        `max_leaf_nodes`:  [2, ) vs [1, )
        
Could you help to clear this up? I am already writing new checks based on the range in the notes, but if it makes more sense to keep the ranges consistent with the existing code, I will update them. Thank you!

glemaitre commented:
It makes sense for `max_leaf_nodes` because if you have 1 split then you automatically have at least 2 leaves.
For the other one, it is less straightforward but if the code makes these checks so these bounds are correct.

genvalen committed:
min_samples_leaf: add tests and validation

genvalen committed:
max_features: add tests and validation

genvalen committed:
min_samples_split: update

genvalen committed:
remove redundant tests

genvalen committed:
max_leaf_nodes: remove redundant tests

genvalen committed:
Merge branch 'BaseDecisionTree_add_check_scalar' of https://github.com/genvalen/scikit-learn into BaseDecisionTree_add_check_scalar

genvalen committed:
Fix lint issue

genvalen committed:
update range for min_weight_fraction_leaf

genvalen committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into BaseDecisionTree_add_check_scalar

genvalen committed:
remove redundant tests

genvalen committed:
min_samples_split: update to include upper bound

genvalen committed:
slight style edit to tests

genvalen committed:
update boundary inclusivity for max_features

genvalen committed:
update order of tests

genvalen committed:
update order of validations to match signature more closely

genvalen commented:
> **Notes on param range in trees** (param: type, range) `max_depth`: int, if not none, then [1, ) `min_samples_split`: int -> [2, number samples], float-> (0, 0.5] `min_samples_leaf`: int -> [1, number of samples], float-> (0, 0.5] `min_weight_fraction_leaf`: float-> [0, 0.5] `max_features`: int -> [1, number of features], float (0, 1], or string (not checked with check_scalar) `max_leaf_nodes`: int, if not none, then [2, ) `min_impurity_decrease`: float-> [0, )
> 
> Edit: updated ranges based on the ranges checked in the code Edit2: note the upper bound for `min_samples_split` and `min_samples_leaf` was not explicitly specified in the code, but I included it in the new checks.

Hi @glemaitre, please review when you are able. Thank you. Also, please note I updated the ranges in the comment above so you can see the ranges I observed in the code

genvalen committed:
remove some boundary args to rely on defaults more and reduce lines

genvalen committed:
ccp_alpha: add tests and validation

genvalen committed:
edit comment

glemaitre commented on sklearn/tree/_classes.py:
```diff
@@ -225,46 +230,66 @@ def fit(self, X, y, sample_weight=None, check_input=True):
+                "min_samples_split",
+                target_type=numbers.Real,
+                min_val=0.0,
+                max_val=0.5,
```
The previous `min_samples_split` was in the range (0, 1] apparently.

```suggestion
                max_val=1.0,
```

glemaitre commented on sklearn/tree/_classes.py:
```diff
@@ -284,36 +309,51 @@ def fit(self, X, y, sample_weight=None, check_input=True):
+                target_type=numbers.Real,
+                min_val=0.0,
+                max_val=1.0,
+                include_boundaries="right",
```
Here there is a subtle ambiguity for `1` that could be an `Integral` or `Real` whether one pass `1.0` or `1`.

Basically, I would expect:

- `None` allows taking all features.
- `1` as an integer will select one feature.
- `1.0` as a real should not be included.

But I am not really sure what is the best here. Maybe @thomasjpfan or @ogrisel remember how we want to make the validation for this ambiguity.

thomasjpfan replied:
My impression was that `1.0` means "take all the features", which is inline with the current implementation.

glemaitre replied:
Looking at different places it seems that we do the same as stated by @thomasjpfan. I am still not a super fan that we can pass both `1` and `1.0` and that they will have different behaviour.

glemaitre commented on sklearn/tree/_classes.py:
```diff
@@ -537,7 +574,7 @@ def _prune_tree(self):
 
         if self.ccp_alpha < 0.0:
-            raise ValueError("ccp_alpha must be greater than or equal to 0")
+            raise ValueError("must be >= 0.0.")
```
do we need this check here?

glemaitre commented on sklearn/tree/_classes.py:
```diff
@@ -537,7 +574,7 @@ def _prune_tree(self):
+            raise ValueError("must be >= 0.0.")
 
         if self.ccp_alpha == 0.0:
             return
```
With your `check_scalar` we will not have this case because we don't include 0.0

glemaitre commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -1984,7 +2025,7 @@ def assert_is_subtree(tree, subtree):
 def test_prune_tree_raises_negative_ccp_alpha():
     clf = DecisionTreeClassifier()
-    msg = "ccp_alpha must be greater than or equal to 0"
+    msg = "must be >= 0.0."
```
since `_prune_tree` is called in `fit` I would expect `check_scalar` to be called before raising this error.

glemaitre commented:
The remaining failure is just an issue linked to regex matching.

genvalen committed:
Update sklearn/tree/_classes.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

genvalen committed:
min_samples_split: update tests

genvalen committed:
ccp_alpha: remove redundant tests

genvalen committed:
update check scalar calls to explicitly reference "name" param

genvalen commented:
I have updated the `check_scalar` calls in this PR based on https://github.com/scikit-learn/scikit-learn/pull/21605#issuecomment-1007486041

thomasjpfan commented on sklearn/tree/_classes.py:
```diff
@@ -225,46 +230,66 @@ def fit(self, X, y, sample_weight=None, check_input=True):
+                self.min_samples_split,
+                name="min_samples_split",
+                target_type=numbers.Integral,
+                min_val=2,
+                max_val=n_samples,
```
Setting the `max_val=n_samples` is breaking out common tests:

https://github.com/scikit-learn/scikit-learn/blob/4769a42be77f9a7b56e191356ffb09a44a49d7b9/sklearn/utils/estimator_checks.py#L1364

where `n_samples=1`. For now, think we set `min_val=1` without an upper bound:

```suggestion
                min_val=1,
```

thomasjpfan reviewed (commented):
For this PR we do need to update the messages in  `test_gbdt_parameter_checks` since we are changing the error messages:

https://github.com/scikit-learn/scikit-learn/blob/4769a42be77f9a7b56e191356ffb09a44a49d7b9/sklearn/ensemble/tests/test_gradient_boosting.py#L119

genvalen committed:
Update sklearn/tree/_classes.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

genvalen committed:
update tests for min_samples_split

genvalen committed:
update messages in test_gbdt_parameter_checks to pass CI

genvalen commented:
> For this PR we do need to update the messages in `test_gbdt_parameter_checks` since we are changing the error messages:
> 
> https://github.com/scikit-learn/scikit-learn/blob/4769a42be77f9a7b56e191356ffb09a44a49d7b9/sklearn/ensemble/tests/test_gradient_boosting.py#L119

Thanks @thomasjpfan, I've added a commit the updates the messages causing the issues.

jjerphan commented on sklearn/ensemble/tests/test_gradient_boosting.py:
```diff
@@ -116,9 +148,9 @@ def test_classification_toy(loss):
+def test_gbdt_parameter_checks(GradientBoosting, X, y, params, err_type, err_msg):
     # Check input parameter validation for GradientBoosting
-    with pytest.raises(ValueError, match=err_msg):
+    with pytest.raises(err_type, match=err_msg):
         GradientBoosting(**params).fit(X, y)
```
```suggestion
    est = GradientBoosting(**params)
    with pytest.raises(err_type, match=err_msg):
        est.fit(X, y)
```

jjerphan reviewed (approved):
LGTM.

genvalen committed:
Update sklearn/ensemble/tests/test_gradient_boosting.py

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

thomasjpfan commented on sklearn/tree/tests/test_tree.py:
```diff
@@ -501,73 +501,6 @@ def test_importances_gini_equal_squared_error():
     assert_array_equal(clf.tree_.n_node_samples, reg.tree_.n_node_samples)
 
 
-def test_max_features():
```
It looks like `test_max_features` got moved down the file, but half the code is the same.  Instead of moving the whole `test_max_features` function over, can we delete the `with pytest.raises(ValueError):` parts from this test and keep everything else in place? By keeping the code in place, it makes it easier to run `git blame` and track history.

In other words, keep the following lines:

https://github.com/scikit-learn/scikit-learn/blob/7f61186cbcbc43cd6ccd717abd097f638b786984/sklearn/tree/tests/test_tree.py#L504-L547

and remove:

https://github.com/scikit-learn/scikit-learn/blob/7f61186cbcbc43cd6ccd717abd097f638b786984/sklearn/tree/tests/test_tree.py#L548-L568

thomasjpfan reviewed (commented):
Minor comment, otherwise this looks good to go.

genvalen committed:
Put test_max_features back on line 504

genvalen committed:
fix lint issue

thomasjpfan commented on sklearn/tree/_classes.py:
```diff
@@ -225,46 +230,65 @@ def fit(self, X, y, sample_weight=None, check_input=True):
+                self.min_samples_split,
+                name="min_samples_split",
+                target_type=numbers.Integral,
+                min_val=1,
```
I was mistaken with my previous comment. This should be 2 to have the same behavior as before

```suggestion
                min_val=2,
```

thomasjpfan commented on sklearn/tree/_classes.py:
```diff
@@ -225,46 +230,65 @@ def fit(self, X, y, sample_weight=None, check_input=True):
+                name="min_samples_leaf",
+                target_type=numbers.Integral,
+                min_val=1,
+                max_val=n_samples,
```
This is technically not backward compatible. The following works to form single node trees:

```python
from sklearn.datasets import make_classification
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeClassifier
X, y = make_classification(random_state=0)
clf = DecisionTreeClassifier(random_state=0, min_samples_leaf=X.shape[0] + 1)

clf.fit(X, y)
```

I'm guessing this could still have some educational value. I think we should remove the upper bound:

```suggestion
```

WDYT @glemaitre ?

glemaitre replied:
Yes we can remove the upper bound limit here.

genvalen committed:
Update sklearn/tree/_classes.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

genvalen committed:
Revert "update tests for min_samples_split"

genvalen committed:
update min_val for min_samples_split

genvalen committed:
Update sklearn/tree/_classes.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

glemaitre reviewed (approved):
LGTM.

ogrisel reviewed (approved):
LGTM as well!

ogrisel commented on sklearn/tree/_classes.py:
```diff
@@ -225,46 +230,64 @@ def fit(self, X, y, sample_weight=None, check_input=True):
+                name="min_samples_leaf",
+                target_type=numbers.Real,
+                min_val=0.0,
+                max_val=0.5,
```
I think that for consistency, we need to remove this upper-bound because the integer parametrization allows for more than `int(0.5 * n_samples)`.

```suggestion
```

genvalen committed:
Update sklearn/tree/_classes.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

genvalen committed:
update tests

ogrisel commented:
The new release of numpydoc-1.2 is making `test_check_docstring_parameters` fail but this is unrelated to this PR.
